File: app/src/main/java/com/jcr/sharedtasks/ui/list/TasksListViewModel.java
package com.jcr.sharedtasks.ui.list;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.VisibleForTesting;

import com.jcr.sharedtasks.model.ProjectReference;
import com.jcr.sharedtasks.model.Task;
import com.jcr.sharedtasks.repository.ProjectsRepository;
import com.jcr.sharedtasks.util.AbsentLiveData;
import com.jcr.sharedtasks.util.DeepLinkUtils;
import com.jcr.sharedtasks.util.Objects;

import java.util.List;

import javax.inject.Inject;

public class TasksListViewModel extends ViewModel {

    @VisibleForTesting
    final MutableLiveData<String> projectUUID;
    private final LiveData<List<Task>> tasks;
    private final ProjectsRepository repository;
    private final LiveData<ProjectReference> projectReference;

    @Inject
    public TasksListViewModel(ProjectsRepository repository) {
        this.projectUUID = new MutableLiveData<>();
        this.repository = repository;
        tasks = Transformations.switchMap(projectUUID, input -> {
            if (input.isEmpty()) {
                return AbsentLiveData.create();
            }
            return repository.loadTasks(input);
        });
        projectReference = Transformations.switchMap(projectUUID, input -> {
            if (input.isEmpty()) {
                return AbsentLiveData.create();
            }
            return repository.getProjectReferenceById(input);
        });
    }

    public String getDeepLinkOfCurrentProject() {
        return DeepLinkUtils.DEEPLINK_HOST + repository.getCurrentProjectName().replace(" ", "&")
                + "/" + repository.getCurrentProjectUUID();
    }

    public void setProjectUUID(String projectUUID) {
        if (Objects.equals(projectUUID, this.projectUUID.getValue())) {
            return;
        }
        this.projectUUID.setValue(projectUUID);
    }

    public LiveData<List<Task>> getTasks() {
        return tasks;
    }

    public LiveData<ProjectReference> getProjectReference() {
        return projectReference;
    }

    public void updateTaskStatus(Task task) {
        this.repository.updateTaskStatus(task);
    }

    public void updateTaskAssignee(Task task) {
        this.repository.updateTaskAssignee(task);
    }
}

File: app/src/main/java/com/jcr/sharedtasks/ui/list/TasksListViewModel.kt
package com.jcr.sharedtasks.ui.list

import androidx.lifecycle.*

import com.jcr.sharedtasks.model.ProjectReference
import com.jcr.sharedtasks.model.Task
import com.jcr.sharedtasks.repository.ProjectsRepository
import com.jcr.sharedtasks.testing.OpenForTesting
import com.jcr.sharedtasks.util.AbsentLiveData
import com.jcr.sharedtasks.util.DeepLinkUtils
import com.jcr.sharedtasks.util.Objects
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

import javax.inject.Inject

@OpenForTesting
class TasksListViewModel @Inject
constructor(private val repository: ProjectsRepository) : ViewModel() {

    private val projectUUID = MutableStateFlow<String?>(null)

    val tasks: StateFlow<List<Task>> = projectUUID.transform { projectUUID ->
        projectUUID?.let {
            emitAll(repository.loadTasks(projectUUID))
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = emptyList()
    )

    val projectReference: StateFlow<ProjectReference?> = projectUUID.transform { projectUUID ->
        projectUUID?.let {
            emitAll(repository.getProjectReferenceById(projectUUID))
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = null
    )


    val deepLinkOfCurrentProject: String
        get() = (DeepLinkUtils.DEEPLINK_HOST + repository.currentProjectName.replace(" ", "&")
                + "/" + repository.currentProjectUUID)

    fun setProjectUUID(projectUUID: String) {
        if (Objects.equals(projectUUID, this.projectUUID.value)) {
            return
        }
        this.projectUUID.value = projectUUID
    }

    fun updateTaskStatus(task: Task) {
        viewModelScope.launch {
            repository.updateTaskState(task)
        }
    }

    fun updateTaskAssignee(task: Task) {
        viewModelScope.launch {
            repository.updateTaskAssignee(task)
        }
    }
}
